### PR TITLE
Add slashing to `MultiStaker`

### PR DESCRIPTION
**Pull request by 0xfuturistic**
This PR also includes tests

### CODE REVIEW

**Comment by josojo** [comment]
I though we would not have slashing, after the comment in https://github.com/0xfuturistic/UniStaker/pull/4#issuecomment-2291847280 - "the argument was that as long as the deposit value only increased".
 I have doubts that having one collateral token + permissionless staking with slashing is a good idea. 

Here is an attack vector(though it requires _isPossessory staking):
I create my own staking contract that _isPossessory and slashes any user by giving all tokens to me. Then I use my own coins to stake to this service and get slashed. After doing it, I can still use my collateral token in lending markets to get some other tokens and run away with them.

Coding wise the PR looks good.

**Comment by 0xfuturistic** [comment]
maybe slashing is possible if possessory stakers count == 0?

**Comment by 0xfuturistic** [comment]
we may have to do _permissioned_ possessory stakers

File: src/MultiStaker.sol [hunk 1]
```diff
@@ -37,6 +37,12 @@ contract MultiStaker {
     /// @notice Emitted when `account` withdraws `value` amount of STAKE_TOKEN from `depositId` deposit.
     event Withdrawn(address indexed account, uint256 indexed depositId, uint256 value);
 
+    /// @notice Emitted when `staker` slashes `account` for  `value` amount of STAKE_TOKEN from `depositId` deposit.
+    event Slashed(address indexed account, address indexed staker, uint256 indexed depositId, uint256 value);
+
+    /// @notice Emitted when `account` slashes `value` amount of STAKE_TOKEN from `depositId` deposit.
```
Comment (josojo) [inline src/MultiStaker.sol:43]: In the description there is value that is no in the event

Comment (josojo) [reply src/MultiStaker.sol:43]: And it seems the event is not used

File: src/example-stakers/UniStakerWrapper.sol [hunk 2]
```diff
@@ -84,12 +84,14 @@ contract UniStakerWrapper is IStaker {
...
-        returns (uint256 _stakeId)
-    {
+    function stake(
+        bool _isPossessory,
+        bool,
+        address,
+        uint256 _value,
+        uint256 _unstakeGasAmount,
+        bytes calldata _extraData
+    ) external onlyMultiStaker matchIsPossessory(_isPossessory) returns (uint256 _stakeId) {
```
Comment (josojo) [inline src/example-stakers/UniStakerWrapper.sol:94]: I am not sure whether we should have a a matchIsSlashable(). It's not always needed, buts still it seems good to have that in general to verify that everything is setup correctly.
I am actually unsure here, what do you think?

**Review by josojo — APPROVED** [review]
Looks good!

why don't we allow for partial slashing? Is there any reason for it?

File: src/example-stakers/UniStakerWrapper.sol (hunk 2 above)
Comment (0xfuturistic) [reply src/example-stakers/UniStakerWrapper.sol:94]: good idea! we could add it to a base contract for stakers

Comment (0xfuturistic) [reply src/example-stakers/UniStakerWrapper.sol:94]: will do that in a different PR

**Pull request merged**